### PR TITLE
Issue #4870: Mark checks with MT interfaces (part 1)

### PR DESCRIPTION
**Pull request by soon**
Issue #4870

This PR marks checks with `stateless / global stateful` or `file-stateful` interfaces.

This PR does not cover all checks, it only includes modules, which does not require additional changes, i.e. only `@FIleStatefulCheck`, `@StatelessCheck` or `@GlobalStatefulCheck'.

The other checks must be changed in the other PRs

### CODE REVIEW

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/header/AbstractHeaderCheck.java [hunk 1]
```diff
@@ -46,7 +47,7 @@
  * @author o_sukhosolsky
  */
 public abstract class AbstractHeaderCheck extends AbstractFileSetCheck
-    implements ExternalResourceHolder {
+    implements ExternalResourceHolder, OneModuleInstancePerApplication {
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/header/AbstractHeaderCheck.java]: Marking abstract classes is a risky business.
`OneModuleInstancePerApplication` is the weakest type to apply to a module. What if a module that implements this abstract class requires the strongest type, like `OneModuleInstancePerThread`.
You will now have a conflict and the module will implement both interfaces.

IMO, we either shouldn't mark abstract classes or if the define both, `OneModuleInstancePerThread` takes precedence.

Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/header/AbstractHeaderCheck.java]: Also why didn't you mark `AbstractCheck` and `AbstractFileSetCheck` but `AbstractJavadocCheck`. Once you mark the base, all checks are now forced under that if it is the strongest type.
If they are marked strong, we need to rethink how they are coded as they will bring down all the other checks.

**Comment by soon** [comment]
> Please expand in issue what are the requirements for a module to be marked with a specific interface.

Updated issue https://github.com/checkstyle/checkstyle/issues/4870

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/header/AbstractHeaderCheck.java (hunk 1 above)
Comment (soon) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/header/AbstractHeaderCheck.java]: @rnveach Changed CheckerCloneService, now `OneModuleInstancePerThread` takes precedence.

Comment (soon) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/header/AbstractHeaderCheck.java]: > Also why didn't you mark AbstractCheck and AbstractFileSetCheck but AbstractJavadocCheck. Once you mark the base, all checks are now forced under that if it is the strongest type.

As far as I remember, there was a problem with `blockCommentAst` field. I think, it is better to move `AbstractJavadocCheck` changes to another PR

**Comment by rnveach** [comment]
> I think, it is better to move AbstractJavadocCheck changes to another PR

Please do this and give it it's own issue if it requires to be thread safe like the other PRs you started.

**Comment by soon** [comment]
> Please do this and give it it's own issue if it requires to be thread safe like the other PRs you started.

@rnveach Done, here is the PR: https://github.com/checkstyle/checkstyle/pull/4926

**Comment by romani** [comment]
we released 8.2 version, please rebase all your PRs to our latest master to avoid CI problems

**Comment by rnveach** [comment]
@soon Since the annotations have been merged, this should be the next PR we concentrate on, right?

If so, please rebase and remove the first commit as it was merged from a separate PR.

**Comment by soon** [comment]
@rnveach Yes, working on it

**Comment by soon** [comment]
@rnveach Moved `TranslationCheck` to another PR: https://github.com/checkstyle/checkstyle/pull/5165 as it requires additional changes. This PR contains only checks without additional changes

**Comment by rnveach** [comment]
@soon ok, since you are moving some checks out of here, we will need to make UT later.
Either as a last thing for this issue or part of another.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportControlCheck.java [hunk 2]
```diff
@@ -44,7 +45,9 @@
  *
  * @author Oliver Burn
  */
-public class ImportControlCheck extends AbstractCheck implements ExternalResourceHolder {
+@FileStatefulCheck
+public class ImportControlCheck extends AbstractCheck
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportControlCheck.java]: please undo line split.

Comment (soon) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportControlCheck.java]: Fixed

**Comment by rnveach** [comment]
@soon Can you rebase on latest master to try and fix wercker.
@romani Do you know what this error is about?
> Dependency convergence error for org.seleniumhq.selenium:selenium-remote-driver:3.6.0 paths to dependency are:

**Comment by rnveach** [comment]
@romani Please review.

**Comment by soon** [comment]
> @soon Can you rebase on latest master to try and fix wercker.

@rnveach Done

**Comment by sabaka** [comment]
@soon , I still can see many line's split. Could you undo these changes?

Please create issue to create UT Richard requested earlier here.

**Comment by soon** [comment]
@sabaka Done: https://github.com/checkstyle/checkstyle/issues/5190

**Comment by rnveach** [comment]
@romani ping

**Pull request merged**